timemachine/cpp/src/observable.cu

```diff
@@ -24,27 +56,61 @@ void AvgPartialUPartialParam::observe(
     double lambda) {
 
     if(step % interval_ == 0) {
+        cudaStream_t stream = static_cast<cudaStream_t>(0);
+        const int size = bp_->size();
+        // Need the latest du_dp, so reset to zero each round
+        gpuErrchk(cudaMemset(d_du_dp_, 0, size*sizeof(*d_du_dp_)));
         bp_->execute_device(
             N,
             d_x_t,
             d_box_t,
             lambda,
             nullptr,
-            d_sum_du_dp_,
+            d_du_dp_,
             nullptr,
             nullptr,
-            static_cast<cudaStream_t>(0) // TBD: parallelize me!
+            stream // TBD: parallelize me!
         );
-        count_ += 1;
+        count_++;
+        const int tpb = 32;
+        const int blocks = (size+tpb-1)/tpb;
+        k_compute_variance<<<blocks, tpb, 0, stream>>>(
+            size,
+            count_,
+            d_du_dp_,
+            d_m_,
+            d_s_
+        );
+
+        gpuErrchk(cudaPeekAtLastError());
     }
 
 }
 
+void AvgPartialUPartialParam::std_du_dp(double *h_buf) const {
+    // Copying is only necessary if there is a variance to return
+    if (count_ > 1) {
+        gpuErrchk(cudaMemcpy(h_buf, d_s_, this->bp_->size()*sizeof(*h_buf), cudaMemcpyDeviceToHost));
+    }
+    for(int i=0; i < this->bp_->size(); i++) {
+        if (count_ <= 1) {
+            h_buf[i] = 0.0;
+        } else {
+            // Population variance to copy default np.std behavior
```

Comment:
This is a difference from the method described in https://www.johndcook.com/blog/standard_deviation/, though the sample variance differs from np.std which seems to be used in our code base

Reply:
Are you referring to the off by 1? i.e. dividing by /count_ as opposed to /(count_-1)?

Reply:
Yes, to replicate the sample variance as defined in the link above it would be (count_ - 1). Which you can accomplish in numpy with `np.std(x, ddof=1)`

Reply:
I learnt something new!